### ISSUE TITLE
Windows x64 "undefined reference" compile errors in files adig.c & ahost.c with commit 9f03635

### ISSUE THREAD

**Issue by jackL999** [c-ares/c-ares#694]
Using x86_64-w64-mingw64-gcc 13.2 on Linux these are just some of the errors I get when trying to compile. 

Errors building c-ares
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:463: adig.exe] Error 1
make[2]: Leaving directory '/home/jjlem/sources/c-ares/build/src/tools'
make[1]: *** [Makefile:407: all-recursive] Error 1
make[1]: Leaving directory '/home/jjlem/sources/c-ares/build/src'
make: *** [Makefile:533: all-recursive] Error 1

usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: ahost-ahost.o: in function `ai_callback':
/home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/ahost.c:262: undefined reference to `__imp_ares_inet_ntop'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/ahost.c:266: undefined reference to `__imp_ares_freeaddrinfo'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/ahost.c:244: undefined reference to `__imp_ares_strerror'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: ahost-ahost.o: in function `callback':
/home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/ahost.c:227: undefined reference to `__imp_ares_inet_ntop'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/ahost.c:223: undefined reference to `__imp_ares_strerror'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: ahost-ahost.o: in function `main':
/home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/ahost.c:92: undefined reference to `__imp_ares_library_init'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/ahost.c:151: undefined reference to `__imp_ares_init_options'

/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/adig.c:281: undefined reference to `__imp_ares_dns_opcode_tostr'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/adig.c:286: undefined reference to `__imp_ares_dns_record_get_flags'
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /home/jjlem/sources/c-ares/build/src/tools/../../.././src/tools/adig.c:291: undefined reference to `__imp_ares_dns_record_rr_cnt'

**Comment by bradh352** [comment]
What build system?  What version of c-ares? What is your environment (msys2, cygwin, etc)?

**Comment by bradh352** [comment]
I should also mention if you're using autotools, you cannot build both shared and static simultaneously as autotools doesn't appear to provide a way to rebuild all the sources to generate both versions of the library (due to the way symbol name mangling occurs).  CMake does, however.

**Comment by jackL999** [comment]
Build system is Arch Linux latest rolling update
autoreconf (GNU Autoconf) 2.72
GNU Make 4.4.1 
x86_64-w64-mingw32-gcc   version 13.1.0
.././configure --build=x86_64-linux-gnu --host=x86_64-w64-mingw32 --prefix=/home/jjlem/sources/win-build --disable-shared --enable-static --without-random LIBS=lws2_32

Doing a git reset --hard 52e72fe back to November 15 the repository version I had before doing  git pull to the latest commit followed by autoreconf -fi and everything compiles perfectly just as it had before.

**Comment by bagder** [comment]
> if you're using autotools, you cannot build both shared and static simultaneously

How did we mess that up? autotools and libtool have built both by default since decades...

**Comment by bradh352** [comment]
I'm using Ubuntu 22.04 with the mingw-w64 package that it contains, which is probably a bit older, but this is the build I used and it worked fine:
```
git clone https://github.com/c-ares/c-ares
cd c-ares
autoreconf -fi
mkdir build
cd build
../configure --host=x86_64-w64-mingw32 --prefix="$HOME/winbuild" --enable-static --disable-shared --disable-tests
make V=1
```

I may need the full output of that sequence from your system, if you can capture the console output from that sequence above and attach it.

**Comment by jackL999** [comment]
From working through the history of the files ahost.c and adig.c I discovered that the commit
cb9ad0e - (HEAD -> main) sonarcloud: const (4 weeks ago) on  Dec 19, 2023 builds without errors but that commit
4f490b4 - (HEAD -> main) Autotools: rework to simplify and fix recent issues (#674) (3 weeks ago) on Dec 28, 2023 fails with the errors noted previously.

**Comment by bradh352** [comment]
> > if you're using autotools, you cannot build both shared and static simultaneously
> 
> How did we mess that up? autotools and libtool have built both by default since decades...

Not on Windows with autotools afaik.

The problem is when building a DLL, the library must be compiled with `__declspec(dllexport)` on all public symbols, and the header for any libraries or tools that link with c-ares must have those public functions as `__declspec(dllimport)` but on static libraries, you MUST NOT have `__declspec` specified at all.  This is all controlled via `CARES_STATICLIB` define which is at compile time.  So to create both shared and static libraries on Windows, you have to compile every file *twice* one with CARES_STATICLIB and one without.

The problem is the symbols,

**Comment by bradh352** [comment]
@jackL999 autotools was completely rewritten, we knew there would be errors in the field.  The release notes for c-ares 1.25.0 says as much and references that commit where you said the breakage occurred.

**Comment by jackL999** [comment]
> @jackL999 autotools was completely rewritten, we knew there would be errors in the field. The release notes for c-ares 1.25.0 says as much and references that commit where you said the breakage occurred.

OK, so that is where the problem happened. Everything before commit 4f490b4 builds completely fine on my system, anything after that doesn't.

**Comment by bagder** [comment]
> one with CARES_STATICLIB and one without

Ah right, that old thing. That's not autotool's fault though, that's a choice in how the build is done, leaning on the define instead of doing it another way. I just never cared, I suppose mostly because it is only for Windows...

**Comment by bradh352** [comment]
> > @jackL999 autotools was completely rewritten, we knew there would be errors in the field. The release notes for c-ares 1.25.0 says as much and references that commit where you said the breakage occurred.
> 
> OK, so that is where the problem happened. Everything before commit [4f490b4](https://github.com/c-ares/c-ares/commit/4f490b40df72beead4d70e267161a7066ff11ea8) builds completely fine on my system, anything after that doesn't.

Right, so if you could provide the output I requested above, I can work on debugging this.

**Comment by jackL999** [comment]
> > > @jackL999 autotools was completely rewritten, we knew there would be errors in the field. The release notes for c-ares 1.25.0 says as much and references that commit where you said the breakage occurred.
> > 
> > 
> > OK, so that is where the problem happened. Everything before commit [4f490b4](https://github.com/c-ares/c-ares/commit/4f490b40df72beead4d70e267161a7066ff11ea8) builds completely fine on my system, anything after that doesn't.
> 
> Right, so if you could provide the output I requested above, I can work on debugging this.

But from commit 4f490b4 where the problem first happens , right?
git reset --hard 4f490b4

**Comment by bradh352** [comment]
> > one with CARES_STATICLIB and one without
> 
> Ah right, that old thing. That's not autotool's fault though, that's a choice in how the build is done, leaning on the define instead of doing it another way. I just never cared, I suppose mostly because it is only for Windows...

Yeah, you might be able to do it with a DEF file.  But really not worth it.  I'm honestly surprised autotools is used for Windows builds :)

**Comment by bradh352** [comment]
@jackL999 yes, that's the commit.  But really, we'd need to know why on the new system its not working but did on the old for you, since it works for me.  Hence needing the output of the command sequence in my comment above I needed the console output of.  Otherwise I don't see how we can fix it if you can't provide that.  I'd strongly recommend not just giving up and reverting, fixing is always best.

**Comment by jackL999** [comment]
bradh352, I got to this point which generated a lot of errors,
>jjlem>sources>
>git clone https://github.com/c-ares/c-ares
Cloning into 'c-ares'...
remote: Enumerating objects: 13605, done.
remote: Counting objects: 100% (830/830), done.
remote: Compressing objects: 100% (176/176), done.
remote: Total 13605 (delta 724), reused 699 (delta 654), pack-reused 12775
Receiving objects: 100% (13605/13605), 6.77 MiB | 17.47 MiB/s, done.
Resolving deltas: 100% (9199/9199), done.
>jjlem>sources> 
>cd c-ares
>jjlem>sources>c-ares>
>autoconf -if
configure.ac:291: warning: Checking for headers with the preprocessor is
configure.ac:291: deprecated. Specify prerequisite code to AC_CHECK_HEADER
configure.ac:291: instead of using fourth argument '-'. (Many headers need
configure.ac:291: no prerequisites. If you truly need to test whether
configure.ac:291: something passes the preprocessor but not the compiler,
configure.ac:291: use AC_PREPROC_IFELSE.)
/usr/share/autoconf/autoconf/headers.m4:115: _AC_CHECK_HEADER_PREPROC is expanded from...
/usr/share/autoconf/autoconf/headers.m4:56: AC_CHECK_HEADER is expanded from...
/usr/share/autoconf/autoconf/headers.m4:217: AC_CHECK_HEADERS is expanded from...
configure.ac:291: the top level
configure.ac:44: error: possibly undefined macro: AM_INIT_AUTOMAKE
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
configure.ac:45: error: possibly undefined macro: AC_ENABLE_SHARED
configure.ac:49: error: possibly undefined macro: AM_PROG_CC_C_O
configure.ac:67: error: possibly undefined macro: AC_MSG_ERROR
configure.ac:108: error: possibly undefined macro: AM_MAINTAINER_MODE
configure.ac:162: error: possibly undefined macro: AM_CONDITIONAL
configure.ac:189: error: possibly undefined macro: AC_DISABLE_STATIC
configure.ac:814: error: possibly undefined macro: AC_MSG_WARN
configure.ac:884: error: possibly undefined macro: AC_CONFIG_FILES
configure.ac:891: error: possibly undefined macro: AM_COND_IF

**Comment by bradh352** [comment]
its autoreconf -fi  not autoconf

**Comment by jackL999** [comment]
> its autoreconf -fi not autoconf

Yes I discovered that myself which is why I deleted the previous reply. But I have found the real error after doing the series of commands you listed.
I was running make -j8 CPPFLAGS="-march=haswell" which is what I do when building many other code libraries and programs, however, with c-ares I found that doesn't work because it likely clobbers the CPPFLAGS elsewhere. So I added CPPFLAGS="-march=haswell" before the configure line as in 

CPPFLAGS="-march=haswell" ../configure --build=x86_64-linux-gnu --host=x86_64-w64-mingw32 --prefix=/home/jjlem/sources/aria2-git/win-build --disable-shared --enable-static --without-random --disable-tests LIBS=-lws2_32

And now everything works including building the libraries from the latest commit. Sorry for wasting your time with my dumb mistakes.
